### DISCUSSION TITLE
How do I execute dagster and its jobs in separate compute environments?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18222] [Q&A] [answered]
Fairly basic question, but I realize I don't really understand how to modify self-hosted Dagster code to move the execution of software-defined materializations out of the hardware running the Dagster instance and into some scalable compute (Kubernetes, Azure Container Apps, Cloud Run, etc...). I think I saw a guide somewhere but can anyone point me to it? Note that I'm on Azure and fairly open minded on how to implement this.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1700671108027109?thread_ts=1700671108.027109&cid=C01U954MEER).*

**Comment by mkleinbort-ic** [comment]
To clarify the ask - this is about scaling "very parallel" asset executions to outside the hardware running Dagster itself. 

Say there is 3 software defined assets as parquet files in blob storage - `X`, `y`, `training_set` - to keep it simple.

And I have many software defined assets that are compute intensive but also independent of each other. 

```python
@asset
def model1(X, y, training_set):
    ... # Magic

@asset
def model2(X, y, training_set):
    ... # Magic

@asset
def model3(X, y, training_set):
    ... # Magic

@asset
def model4(X, y, training_set):
    ... # Magic

@asset
def model5(X, y, training_set):
    ... # Magic
```

What is the best way to get each of these assets to "fan out" and leverage scalable compute (Kubernetes, Lambdas, whatever can run a docker image...) to execute mostly in parallel. 

Or more generally, how to achieve this with an entire Dagster deployment.

**Comment by gibsondan** [comment, marked as answer]
There are two main possibilities in Dagster today:

Dagster Pipes (https://docs.dagster.io/guides/dagster-pipes) lets you call out to external execution environments (k8s / lambda / others) within the body of your assets and receive results back, without having to totally rewrite that compute to be a Dagster asset: 

Executors (https://docs.dagster.io/deployment/executors#executors) let you customize how each asset or op within a given job is executed. If you want each of your assets within a job to run in its own k8s pod with different resources or other configuration, check out the k8s_job_executor, which runs each asset in its own pod. See https://docs.dagster.io/deployment/guides/kubernetes/customizing-your-deployment#kubernetes-configuration-on-individual-steps-in-a-run for an example of customizing the resource requirements for a given asset so that it has access to more memory or CPU than the rest of the assets in the job.
